Src/Base/AMReX_MPMD.cpp

```diff
@@ -245,6 +264,124 @@ Copier::Copier (BoxArray const& ba, DistributionMapping const& dm)
     }
 }
 
+Copier::Copier ()
+{
+    int rank_offset = myproc - ParallelDescriptor::MyProc();
+    int this_root, other_root;
+    if (rank_offset == 0) { // First program
+        this_root = 0;
+        other_root = ParallelDescriptor::NProcs();
+    } else {
+        this_root = rank_offset;
+        other_root = 0;
+    }
+
+    int this_nboxes;
+
+    if (myproc == this_root) {
+        if (rank_offset == 0) // the first program
+        {
+            MPI_Recv(&this_nboxes, 1, MPI_INT, other_root, 1, MPI_COMM_WORLD,
+                     MPI_STATUS_IGNORE);
+            m_bv.resize(this_nboxes);
+            MPI_Recv(m_bv.data(), this_nboxes,
+                     ParallelDescriptor::Mpi_typemap<Box>::type(),
+                     other_root, 3, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+        }
+        else // the second program
+        {
+            MPI_Recv(&this_nboxes, 1, MPI_INT, other_root, 0, MPI_COMM_WORLD,
+                     MPI_STATUS_IGNORE);
+            m_bv.resize(this_nboxes);
+            MPI_Recv(m_bv.data(), this_nboxes,
+                     ParallelDescriptor::Mpi_typemap<Box>::type(),
+                     other_root, 2, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+        }
+    }
+
+    ParallelDescriptor::Bcast(&this_nboxes, 1);
+    if (m_bv.empty()) {
+        m_bv.resize(this_nboxes);
+    }
+
+    ParallelDescriptor::Bcast(m_bv.data(), m_bv.size());
+    Vector<Box> bv(m_bv);
+    BoxArray ba(BoxList(std::move(bv)));
+    m_dm.define(ba);
+    Vector<int> procs = m_dm.ProcessorMap();
+    if (rank_offset != 0) {
+        for (int i = 0; i < this_nboxes; ++i) {
+            procs[i] += rank_offset;
+        }
+    }
+
+    Vector<int> oprocs;
+    if (myproc == this_root) {
+        if (rank_offset == 0) // the first program
+        {
+            MPI_Send(procs.data(), this_nboxes, MPI_INT, other_root, 4, MPI_COMM_WORLD);
+            oprocs.resize(this_nboxes);
+            MPI_Recv(oprocs.data(), this_nboxes, MPI_INT, other_root, 5, MPI_COMM_WORLD,
+                     MPI_STATUS_IGNORE);
+        }
+        else // the second program
+        {
+            oprocs.resize(this_nboxes);
+            MPI_Recv(oprocs.data(), this_nboxes, MPI_INT, other_root, 4, MPI_COMM_WORLD,
+                     MPI_STATUS_IGNORE);
+            MPI_Send(procs.data(), this_nboxes, MPI_INT, other_root, 5, MPI_COMM_WORLD);
+        }
+    }
+
+    if (oprocs.empty()) {
+        oprocs.resize(this_nboxes);
+    }
+
+    ParallelDescriptor::Bcast(oprocs.data(), oprocs.size());
+
+    // procs holds mpi ranks of our boxes, and oprocs holds
+    // mpi ranks of the other program's boxes.  All mpi ranks are in
+    // MPI_COMM_WORLD.
+
+    // Build communication meta-data
+    m_is_thread_safe = ba.ixType().cellCentered();
+
+    std::vector<std::pair<int,Box> > isects;
+
+    for (int i = 0; i < this_nboxes; ++i) {
+        if (procs[i] == myproc) {
+            ba.intersections(m_bv[i], isects);
```

Comment:
No need to use BoxArray::intersections. because this is a special case that we copy from box i to box i, not to box j where j!=i.

Reply:
Thanks for helping me with this, @WeiqunZhang ! For this special case if we ONLY do a copy of a (NODAL) box `i` to box `i` do you think we should also make `m_is_thread_safe = true` irrespective of the `BoxArray` being cell centered or not?